### DISCUSSION TITLE
Question: How to install Khoj in Windows?

### DISCUSSION THREAD

**Discussion by Liong1976** [khoj-ai/khoj#154] [Q&A] [answered]
Hi,

First of all, thank you for making this awesome plugin.

FYI, I am not a coder and know nothing about computer programming. 

How to install this plugin in Windows, especially in this part? Could you please give me a step-by-step guide? 

![image](https://user-images.githubusercontent.com/88898230/218265838-3bcc8eec-9588-4971-8966-a3e9fcb485fb.png)

I appreciate your help.

**Comment by debanjum** [comment, marked as answer]
Hi @Liong1976, if you want to use the khoj obsidian plugin follow [these instructions](https://github.com/debanjum/khoj/tree/master/src/interface/obsidian#Setup). 
For more detailed instructions, see below:
1. Open [windows command prompt](https://www.makeuseof.com/tag/a-beginners-guide-to-the-windows-command-line/) by searching for "command prompt" from your windows start menu
2. Paste the following command in the command prompt: `pip install khoj-assistant && khoj --no-gui` and hit Enter
3. Wait until you see something like below screenshot in your command prompt
    <img width="581" alt="Screenshot 2023-02-11 at 10 09 41" src="https://user-images.githubusercontent.com/6413477/218268526-cedbe1bd-ad5c-436b-af4b-2e8af88852e8.png">
4. Now (re-)Install the Khoj Obsidian Plugin and wait ~5 minutes (while khoj indexes your notes)
5. Finally follow [these instructions](https://github.com/debanjum/khoj/tree/master/src/interface/obsidian#use) to use Khoj from Obsidian

P.S: I'm moving this issue into a q&a discussion

**Reply by jasmeetchadha** [reply]
I get an error 

'pip' is not recognized as an internal or external command,
operable program or batch file.

It would be really great if you could share detailed instructions for non-coders.

**Reply by Liong1976** [reply]
Thanks, @debanjum 
Unfortunately, I got the same problem as @jasmeetchadha did.

![image](https://user-images.githubusercontent.com/88898230/218286415-7bd42c2a-0b2f-4623-a76a-e138cfcf3855.png)

**Reply by jasmeetchadha** [reply]
Go to cmd prompt and write python and enter. It will install it from Microsoft store… and then the pip command will work. 

My issue now is that after installing using pip, the backend still doesn’t work

**Reply by debanjum** [reply]
> I get an error
> 
> 'pip' is not recognized as an internal or external command, operable program or batch file.
> 
> It would be really great if you could share detailed instructions for non-coders.

Yeah, the instructions and installation process hasn't been tuned for non-coders yet. I'll update them soon based on feedback like this. But until then opening discussion topics here and looking up ways to fix errors on the internet are the best way to get stuff working

**Reply by Liong1976** [reply]
Hi @debanjum,

This is what I got after I run `pip install khoj-assistant && khoj --no-gui`

![image](https://user-images.githubusercontent.com/88898230/218315483-ef3160dd-67f9-4be0-8668-1aed048041b7.png)

After I used this `pip install khoj-assistant`, this is what I got

![image](https://user-images.githubusercontent.com/88898230/218315563-d71ff8c8-f1df-4bef-b84e-9094f31ed1f3.png)

**Reply by debanjum** [reply]
Strange, not sure why the torch package is invalid. Maybe it got corrupted? Can you try the steps below:
 
1. Try install the khoj backend again but with this command 
   ```shell
   pip install --no-cache-dir --upgrade khoj-assistant 
   ```

If that doesn't work
1. Try install the `torch` packages separately first and then install the khoj backend by running the following command sin your cmd prompt:
    ```shell
    pip install torch torchvision torchaudio
    pip install --upgrade khoj-assistant
    ```

**Reply by Liong1976** [reply]
Thanks, @debanjum 

This is what I got after I run 

`pip install --no-cache-dir --upgrade khoj-assistant` 

![image](https://user-images.githubusercontent.com/88898230/218341944-e8e43198-974e-4e2b-ba87-89b128235aaa.png)

This is what I got after running 

`pip install torch torchvision torchaudio`

![image](https://user-images.githubusercontent.com/88898230/218342043-07e25ffa-296a-4deb-a6e5-3396ddcae344.png)

**Reply by debanjum** [reply]
Ok, so this is even stranger and the message looks scary. But looking at 
1. How the downloads for both `pydantic` and `torch` packages did not complete (e.g `0.0 / 2.0Mb` and `9.2/162.6 Mb`) 
2. At similar errors online. E.g [1](https://stackoverflow.com/questions/71435874/pip-these-packages-do-mot-match-the-hashes-from-the-requirements-file), [2](https://stackoverflow.com/questions/59823392/error-these-packages-do-not-match-the-hashes-from-the-requirements-file-when-u)

It seems like it could be an issue with flaky internet.

A few things to try:
1. If you can, jump to a stable internet connection before you try install the khoj backend
2. Ensure your computer isn't on an internet running behind a proxy server or some such (e.g at a university)
4. Try the `pip install --no-cache-dir --upgrade khoj-assistant`  a few times.

**Reply by Liong1976** [reply]
Aha, I think that's the problem. I am using the internet from a university. Unfortunately, I don't have an internet connection on my own. I only can use public internets such as from a university, a public library, or a coffee shop.

**Reply by debanjum** [reply]
Yeah, that's unfortunate as public internet can be throttled and flaky. Best way maybe to put stuff to download during off peak hours at university, library or coffee shop. Pip doesn't retry with partial downloads [yet](https://github.com/pypa/pip/issues/4796)

A few ways to workaround this:
- Try a few times to download khoj with `pip install --upgrade khoj-assistant` but manually delete corrupted packages between retries
- Install khoj dependencies separately. E.g `pip install --no-cache-dir torch` until download completes successfully 
- Download wheels for khoj dependencies like [torch](https://files.pythonhosted.org/packages/33/bd/e174e6737daba03f8eaa7c051b9971d361022eb37b86cbe5db0b08cab00e/torch-1.13.1-cp310-cp310-win_amd64.whl), [pydantic](https://files.pythonhosted.org/packages/68/17/1828a267f999077fba10b24076e027df79cc2c5ede8e0a91a7cbf43428d8/pydantic-1.9.1-cp310-cp310-win_amd64.whl) etc separately directly from pypi via browser. 
  You find the right `whl` to download for your system looking at version in [khoj setup.py](https://github.com/debanjum/khoj/blob/master/setup.py) with `cp310` (for python 3.10) and `win_amd64` (for windows 64bit) in name of `.whl` file listed in `https://pypi.org/project/[name-of-package]/[version-from-khoj-setup.py]/#files`

**Reply by Liong1976** [reply]
Hi @debanjum,  

Thanks for the answer

>Try a few times to download khoj with pip install --upgrade khoj-assistant but manually delete corrupted packages between retries

Can you please let me know how to delete the corrupted files? Where are the files located? This way is the most doable for me since I don't know how to do the others.

**Reply by debanjum** [reply]
Actually, just doing a combination of the first and second workarounds suggested. 
This should avoid having to delete the corrupted packages manually. That is,
1. Run `pip install --upgrade --pre khoj-assistant`
2. When a package fails to install, run `pip install --no-cache-dir <name-of-failed-package>` until failed package installed successfully. E.g `pip install --no-cache-dir torch`
3. Once the failed package has installed, re-run `pip install --upgrade --pre khoj-assistant`. This should continue installing the remaining packages required by khoj

Repeat steps 2. and 3., until the `khoj-assistant` package is successfully installed

**Reply by Liong1976** [reply]
Thanks, @debanjum 

I'll try your suggestion

**Reply by Liong1976** [reply]
Hi @debanjum,

I managed to install `khoj-assistant`

![image](https://user-images.githubusercontent.com/88898230/218802694-75723fc8-41f5-4dda-90dc-2de36782dc2e.png)

Unfortunately, your plugin is still not working.

![image](https://user-images.githubusercontent.com/88898230/218803667-27ea0dea-ace9-4312-84a2-2e86caeddc43.png)

![image](https://user-images.githubusercontent.com/88898230/218804078-d97ce35f-114e-40d1-8f58-df5c86cf98b4.png)

Please let me know how to solve the problem. Thanks

**Reply by debanjum** [reply]
> I managed to install `khoj-assistant`

Good to know you were finally able to install it! 🎉 

> Unfortunately, your plugin is still not working.

- Did you start the khoj backend with `khoj --no-gui -vv` in your cmd terminal?
  This is why you see the *Khoj backend disconnected* errors when you install the khoj plugin in obsidian.

- Also the khoj backend is required to be running **before** you install/enable the khoj plugin in obsidian. 
  This allows the plugin to configure the khoj backend to index your obsidian vault

**Reply by Liong1976** [reply]
This is what I got after running `khoj --no-gui -vv`

![image](https://user-images.githubusercontent.com/88898230/218818200-96369313-38ad-4168-b573-41daef2ced35.png)

**Reply by debanjum** [reply]
This is because the directory in which khoj is installed is not in your `PATH`. (Other windows users had a similar problem). 
Try run `C:\Users\%USER%\AppData\Local\Programs\Python\Python310\Scripts\khoj.exe --no-gui` to see if that starts khoj for you

**Reply by Liong1976** [reply]
I couldn't locate the file that you mentioned.

![image](https://user-images.githubusercontent.com/88898230/218821014-5db98a9d-96da-470c-b789-7c7e597e6815.png)

**Reply by debanjum** [reply]
Based on the last `WARNING` in your khoj installation logs
![](https://user-images.githubusercontent.com/88898230/218802694-75723fc8-41f5-4dda-90dc-2de36782dc2e.png)

It seems like the khoj.exe script is installed at `C:\Users\panch\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts\`. 

So maybe try `C:\Users\panch\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts\khoj.exe --no-gui -vv` to start khoj server?

**Reply by Liong1976** [reply]
Thanks, @debanjum 

This is what I got

![image](https://user-images.githubusercontent.com/88898230/218824234-0eb7d800-6302-461e-98da-3add268f73be.png)

![image](https://user-images.githubusercontent.com/88898230/218824276-2d8de33c-2006-4bea-8421-38c6d6bec981.png)

Should I enable the plugin now?

**Reply by Liong1976** [reply]
I enabled the plugin.

I think it's working.

![image](https://user-images.githubusercontent.com/88898230/218825430-d53abb2a-fb22-48ce-9730-7954c07b1e1b.png)

After finishing the update, can I close Command Prompt and this?

![image](https://user-images.githubusercontent.com/88898230/218826028-3d123481-8e3d-484c-8b0a-cfd628174844.png)


Should I run  `--no-gui -vv` everytime I want to use Khoj?

**Reply by Liong1976** [reply]
I typed a word that was used a lot in my vault in `Search in Khoj,` but it showed nothing

**Reply by debanjum** [reply]
> After finishing the update, can I close Command Prompt and this?
> Should I run --no-gui -vv everytime I want to use Khoj?

Just start the khoj backend with `khoj.exe --no-gui -vv` flags set. This will not show you the khoj desktop GUI. 
But you need the khoj backend to be running in the Command Prompt all the time. This is what allows the khoj obsidian plugin to be able to search (and update the index).

> I enabled the plugin.
I think it's working.

Yeah, so the plugin is able to connect to your khoj backend server running in the Command Prompt. Wait till the indexing is done. And assuming it succeeded without issues. Try search with khoj from obsidian. If it works, then you're all set!

**Reply by debanjum** [reply]
> I typed a word that was used a lot in my vault in `Search in Khoj,` but it showed nothing

Did the indexing in the khoj server finish? Wait for that and ensure there were no errors in the indexing process. After that try the search to see if it works for you

**Reply by Liong1976** [reply]
I typed `born` it showed nothing

![image](https://user-images.githubusercontent.com/88898230/218828237-bd1792f8-b445-42f3-bfc4-b1e6cd6bb8bc.png)

![image](https://user-images.githubusercontent.com/88898230/218828309-09f74ab9-3564-4cb6-b326-51e2f161e4b9.png)

**Reply by Liong1976** [reply]
> > I typed a word that was used a lot in my vault in `Search in Khoj,` but it showed nothing
> 
> Did the indexing in the khoj server finish? Wait for that and ensure there were no errors in the indexing process. After that try the search to see if it works for you

How did I know that the indexing process was finished? 

I did the manual indexing, and it is finished.

![image](https://user-images.githubusercontent.com/88898230/218828990-fdcbe7d2-70aa-4ea2-8a91-9a428f65f621.png)

**Reply by debanjum** [reply]
> I did the manual indexing, and it is finished.

Yeah, once the indexing is done the `Update` button should become enabled/colored again.

But can you share a screenshot of what the backend shows after you hit the "Update" button in the Khoj Obsidian plugin settings?
Or after you restart the khoj server? If there were any errors while indexing your notes, the khoj backend should show red colored error message(s)

**Reply by Liong1976** [reply]
I saw a red message when running the backend

![image](https://user-images.githubusercontent.com/88898230/218833608-938ab493-c94c-41e8-bb24-3762c1a13e86.png)

**Reply by debanjum** [reply]
Ah, unfortunate but that error during indexing seems to be the reason you're not seeing any search results. 🙈
Can you share which version of the khoj backend you have installed by running `khoj -V` in your Command Prompt? 

I'd pushed fixes for this issue https://github.com/debanjum/khoj/issues/89 yesterday with https://github.com/debanjum/khoj/commit/11517ba8eb3a968f1100caaedd3043d9310cf230.

See [this reply](https://github.com/debanjum/khoj/issues/145#issuecomment-1427226561) to delete your `.khoj` directory at `C:\Users\%USER%\.khoj`, update and restart khoj backend and re-enable the khoj obsidian plugin. A few other users who had reported this issue said the latest khoj fixed it. Hopefully it fixes it for you too 🤞🏾

**Reply by Liong1976** [reply]
Here is what I got when I run `khoj -v`

![image](https://user-images.githubusercontent.com/88898230/218839873-aa6d7e09-364c-4413-a4be-e3d161c44553.png)

Just installed Khoj Backend this morning, so I should have the latest version

**Reply by Liong1976** [reply]
Should I delete the folder `.khoj` or the contents of the folder?

![image](https://user-images.githubusercontent.com/88898230/218841010-c3bd82f6-48f6-4a23-9543-bbb1f257ff64.png)

**Reply by debanjum** [reply]
> Here is what I got when I run `khoj -v`

Correction you should run the below command with a capital `V`
`C:\Users\panch\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts\khoj.exe -V`

This will give the version of khoj installed on your machine

**Reply by Liong1976** [reply]
I couldn't run the `khoj.exe` 
Every time I run it, the window only opens a second and closes. Even I couldn't screenshot it.

**Reply by debanjum** [reply]
Oh, what changed? You were able to run the khoj.exe file from the Command Prompt earlier, right?

**Reply by Liong1976** [reply]
Yes, I did. I don't know what changed.

**Reply by debanjum** [reply]
Hmm, not sure. Maybe delete the `.khoj` folder and see if restarting the khoj server with `C:\Users\panch\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts\khoj.exe --no-gui -vv` works?

**Reply by Liong1976** [reply]
Last time I managed to type `--no-gui`, but this time I didn't have time even to type anything, and the windows closed. There were a bunch of words, but I could read them because it was so fast being closed.

**Reply by Liong1976** [reply]
This time I was able to screenshot the words before it closed

![image](https://user-images.githubusercontent.com/88898230/218850911-e4c857db-3ea5-462c-84cb-67d5e2cdd9ba.png)

**Reply by debanjum** [reply]
> Last time I managed to type --no-gui, but this time I didn't have time even to type anything, and the windows closed.

😲. Wait I don't get it. Is the command prompt itself not staying open? What do you mean you didn't have time to type `--no-gui`?
What you need to do is:
1. Close all previous Command Prompt terminals
2. Delete the `C:\Users\panch\.khoj` folder (not just the contents of the folder, the whole `.khoj` folder itself)
3. Open a new Command Prompt terminal in Windows
4. Type `C:\Users\panch\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts\khoj.exe --no-gui -vv` and hit Enter
5. Share screenshot(s) of what you see

**Reply by Liong1976** [reply]
Here is what I got

![image](https://user-images.githubusercontent.com/88898230/218873373-c0f9debc-7d3f-4403-b2d1-1e69760b90d9.png)

**Reply by debanjum** [reply]
Ok, this looks much better. 
1. Can you share what version of khoj you have installed by running `C:\Users\panch\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts\khoj.exe -V` in a new Command Prompt
2. Can you run `pip install --upgrade numpy` to upgrade to the [latest numpy](https://pypi.org/project/numpy/) version (`1.24.2`)
3. Restart the khoj server and see if the `UserWarning: Failed to initialize Numpy` warning goes away?
4. Now if you reinstall/re-enable the khoj obsidian plugin, what do the the khoj server logs show?

**Reply by Liong1976** [reply]
Here is what I got about the version

![image](https://user-images.githubusercontent.com/88898230/218882313-f7fad058-b489-4631-8eb6-b7bf7c18afd7.png)

**Reply by Liong1976** [reply]
Here is what I got after upgrading numpy to the latest version (1.24.2)

![image](https://user-images.githubusercontent.com/88898230/218882784-8aef61f8-b9e1-4819-a83e-b8d7b8537ffb.png)

**Reply by Liong1976** [reply]
Here is what looked like after I re-enabled Khoj Obsidian

![image](https://user-images.githubusercontent.com/88898230/218883289-8fd12eb4-b9c3-428e-8680-b64a91e7f7a7.png)

**Reply by debanjum** [reply]
> Here is what I got about the version
> 
> ![image](https://user-images.githubusercontent.com/88898230/218882313-f7fad058-b489-4631-8eb6-b7bf7c18afd7.png)

Yeah, 0.2.6 is an older version of khoj that doesn't have the fix for the `UnicodeDecodeError`. Follow below steps to fix this:
1. Stop the currently running khoj server/ backend in your Command Prompt, by hitting (Ctrl+C or closing Command Prompt)
2. Run `pip install khoj-assistant==0.3.0a1676404462` in Command Prompt to install the fixed up khoj python package 
3. Delete the `C:\Users\panch\.khoj` folder (not just the contents of the folder, the whole `.khoj` folder itself)
4. Open a new Command Prompt terminal in Windows
5. Type C:\Users\panch\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts\khoj.exe --no-gui -vv
6. Now reinstall/re-enable the khoj obsidian plugin and wait until indexing finishes
7. If no errors, search using khoj from obsidian to see if it works

Let me know if that fixes your issue and/or what the the khoj server logs once you've gone through above steps?

**Reply by debanjum** [reply]
> Here is what I got after upgrading numpy to the latest version (1.24.2)
> 
> ![image](https://user-images.githubusercontent.com/88898230/218882784-8aef61f8-b9e1-4819-a83e-b8d7b8537ffb.png)

Ah, great upgrading numpy seems to have fixed the UserWarning about numpy

**Reply by Liong1976** [reply]
Yeay, finally, it works!  Thank you so much @debanjum for your patience! I do appreciate it.

**Reply by debanjum** [reply]
That is awesome! Your perseverance to get khoj working finally paid off 💫. Hope the tool proves useful for you 🙏

**Reply by Liong1976** [reply]
For me, the most beneficial is `Find Similar Notes`. I haven't found that `Search` with Khoj so helpful.  

Another thing, I still can't open the Notes showing in the result. I pressed `Enter` and also click the notes. But it can't open the notes.

**Reply by debanjum** [reply]
> For me, the most beneficial is Find Similar Notes. I haven't found that Search with Khoj so helpful.

Oh interesting. Is that because the search results aren't relevant? If so, some things to try improve that:
- Hit `Ctrl+Enter` after you've entered your query in the search modal. This should show more relevant results first.
- Ensure you're using longer, natural language queries. So use "the name of my dog" rather than "dog". 
- If you want to search for entries that include/exclude a specific word in it, you can add [word filters](https://github.com/debanjum/khoj#Query-filters) to your query. E.g `my favorite animal +"dog"` or `my favorite animal -"slug"`

> Another thing, I still can't open the Notes showing in the result. I pressed Enter and also click the notes. But it can't open the notes.

Oh thanks for letting me know. That sounds like a bug in the code. Let me look into what could be causing it.

**Reply by Liong1976** [reply]
Thanks for the advice. 

You are welcome!

**Comment by jasmeetchadha** [comment]
ok, so I installed python and pip command installed the backend... but seems backend still doesnt work. And khoj url (http://localhost:8000) doesnt open anything

**Reply by debanjum** [reply]
Based on your other comments, it seems like the khoj backend hasn't been installed yet. It'd be great if you can share screenshots of what you see after you've run the command to start the khoj backend in the cmd prompt

**Comment by jasmeetchadha** [comment]
![image](https://user-images.githubusercontent.com/49923268/218285435-97852c97-edac-4709-8c8e-a479d9ed9e4c.png)

**Reply by debanjum** [reply]
Thanks for sharing a screenshot of the commands you ran. The correct commands to install and run the khoj backend are:
1. `pip install khoj-assistant`. This should install the khoj backend
2. `khoj --no-gui -vv`. This should start the khoj backend. 
3. Wait until you see this in the cmd prompt:
   <img width="581" alt="Screenshot 2023-02-11 at 10 09 41" src="https://user-images.githubusercontent.com/6413477/218268526-cedbe1bd-ad5c-436b-af4b-2e8af88852e8.png">
4. Now (re-)Install the Khoj Obsidian Plugin and wait ~5 minutes (while khoj indexes your notes)
5. Finally follow [these instructions](https://github.com/debanjum/khoj/tree/master/src/interface/obsidian#use) to use Khoj from Obsidian

This is based of [the previous setup steps](https://github.com/debanjum/khoj/discussions/154#discussioncomment-4944175) I'd provided with minor variations (i.e split install and start steps, and start the khoj backend in verbose mode for troubleshooting)

**Reply by jasmeetchadha** [reply]
I can only complete step 1. This is done but step 2, gives an error (so the backend isnt running)
![image](https://user-images.githubusercontent.com/49923268/218317168-ee238d97-fe41-4069-8ee3-2e5d01b4bf48.png)

**Reply by jasmeetchadha** [reply]
So I managed to get the backend running. But the search doesnt yield any results. 

Then I tried this a few times (upgraded numpy) and now the backed is not running again. Although it somehow loads md to dictionaries

![image](https://user-images.githubusercontent.com/49923268/218322433-99752fcc-801a-4df6-b600-31c788e8fa98.png)

**Reply by debanjum** [reply]
Thanks for sharing screenshots! It speeds up troubleshooting the issue

> I can only complete step 1. This is done but step 2, gives an error (so the backend isnt running)

Great that you were able to install the khoj backend. 
Did the numpy upgrade (with `pip install --upgrade numpy`) fix your issue with installing the khoj backend?

> So I managed to get the backend running. But the search doesnt yield any results.

What command did you run to finally start the khoj backend?

To try resolve the UnicodeEncode error, can you:
   1. Delete the khoj folder at `C:\Users\jasme\.khoj`
   2. Restart the khoj backend again

**Reply by jasmeetchadha** [reply]
1) Upgrading numpy fixed the issue with installing khoj backend. 
2) To get the backend running I went to find the khoj.exe and ran it there... Then I added the folder to PATH
3) Currently backend has STOPPED RUNNING AGAIN (deleted folder .khoj)... right now Khoj not configured error. 

somehow I managed to start backend running... but search doesnt work. I have a bunch of errors. Not sure where to start 

![image](https://user-images.githubusercontent.com/49923268/218330366-a1d02bcb-5644-4b3b-a7d9-3372ffca5652.png)

![image](https://user-images.githubusercontent.com/49923268/218330408-caa9e6dc-0eca-417f-887f-85c77a1c540b.png)

![image](https://user-images.githubusercontent.com/49923268/218330380-8858d679-fbfd-49c1-819d-b086195c9114.png)

![image](https://user-images.githubusercontent.com/49923268/218330397-f445c72c-e63d-49e7-b8c8-626fea54eca3.png)

![image](https://user-images.githubusercontent.com/49923268/218330421-3083b79b-90e3-47c1-a08e-8d482f08ff11.png)

![image](https://user-images.githubusercontent.com/49923268/218330444-30c37fbd-f38e-4ca2-a34e-f9bcfb6c7424.png)

**Reply by debanjum** [reply]
> 1. Upgrading numpy fixed the issue with installing khoj backend.

Good to know upgrading numpy resolved installing the khoj backend

>     2. To get the backend running I went to find the khoj.exe and ran it there... Then I added the folder to PATH

I see, that makes sense. Thanks for providing that information. Maybe useful to other users on Windows as well
 
>     3. Currently backend has STOPPED RUNNING AGAIN (deleted folder .khoj)... right now Khoj not configured error.

The *Khoj is not configured* issue is a warning. It can be resolved by reinstall/re-enabling the khoj obsidian plugin. 
Either of those options will result in the khoj obsidian plugin automatically configuring the khoj backend.

> 
> somehow I managed to start backend running... but search doesnt work. I have a bunch of errors. Not sure where to start

Great to know that you're able to get the khoj backend finally running! 

The main issue that you still seem to be facing is the `charmap codec can't encode character` error. 
This is causing the indexing of your markdown files to fail. Which is why search is returning no results. 

I think I've identified (and fixed) the bug causing the error you were seeing in https://github.com/debanjum/khoj/commit/11517ba8eb3a968f1100caaedd3043d9310cf230. 
To verify that it works, can you follow the steps below:
1. Delete the khoj folder at C:\Users\jasme\.khoj
2. Upgrade to the latest pre-release version of khoj backend by running `pip install --upgrade --pre khoj-assistant` in your cmd prompt
3. Restart the khoj backend
4. Re-enable/re-install your Khoj obsidian plugin

**Reply by jasmeetchadha** [reply]
And it finally works!!!! yay

I do have another question/suggestion. Do you think you will be evolving this towards a "question-answering chatbot" which answers from my notes at any point?

**Reply by debanjum** [reply]
> And it finally works!!!! yay

Yay, indeed!

> I do have another question/suggestion. Do you think you will be evolving this towards a "question-answering chatbot" which answers from my notes at any point?

Yup, that's exactly what I'll be working towards next! There's a preliminary version of this on the khoj web interface. See [chat with notes](https://github.com/debanjum/khoj#chat-with-notes) for details

**Reply by jasmeetchadha** [reply]
Thanks! look forward to that integration. 

I am not in the woods with coding but would it at all be possible to do the question/answering and chat on the desktop itself? Ie running inference locally on my cpu/GPU? 

There are two benefits 
1) cost  
2) privacy

**Reply by debanjum** [reply]
> Would it at all be possible to do the question/answering and chat on the desktop itself? Ie running inference locally on my cpu/GPU?

Yeah, I have to still evaluate if an offline model can perform to a satisfactory level for question answering. I'll try include the option to choose between an offline and online model for chat. 

Some pros for using online models (e.g GPT):
1. I expect GPT type models to be much better at chat than offline models. (For search, offline models perform surprisingly well)
2. While everyone's spending appetite is different, the cost of just using GPT for chat isn't actually that much now
3. Online models or services make access via phones easy. Being able to [search, chat etc from a phone](https://github.com/debanjum/khoj#access-khoj-on-mobile) is really useful.

Some cons:
1. Privacy of course, is an unavoidable tradeoff if GPT is chosen for chat
2. The ability to search, chat etc offline is great for reliability. It's not fun to have an integral part of your thinking go completely offline (e.g on a plane or in the mountains). Having some kind of graceful degradation for chat in offline mode could reduce this pain point.

**Reply by jasmeetchadha** [reply]
agree with the tradeoffs. Personally I dont mind the limitations of offline model. 

If you dont mind, I would have a few questions w.r.t. deployment of these models and how they work online vs offline. Would you mind chatting about them on phone or messaging? Not sure how to do it over github.

**Reply by debanjum** [reply]
> agree with the tradeoffs. Personally I dont mind the limitations of offline model.

The option to configure this should allow you to choose 🙂

> If you dont mind, I would have a few questions w.r.t. deployment of these models and how they work online vs offline. Would you mind chatting about them on phone or messaging? Not sure how to do it over github.

Sure, don't mind. Maybe DM me on Twitter? My username on it is same as my github username

**Reply by jasmeetchadha** [reply]
Super, sent you a follow request there
